src/inference/hmc.jl

```diff
@@ -663,16 +502,101 @@ observe(spl::Sampler{<:Hamiltonian},
 #### Default HMC stepsize and mass matrix adaptor
 ####
 
-function AHMCAdaptor(alg::AdaptiveHamiltonian; init_ϵ=alg.init_ϵ)
-    p = AHMC.Preconditioner(getmetricT(alg))
-    nda = AHMC.NesterovDualAveraging(alg.δ, init_ϵ)
-    if getmetricT(alg) == AHMC.UnitEuclideanMetric
-        adaptor = AHMC.NaiveHMCAdaptor(p, nda)
+function AHMCAdaptor(alg::AdaptiveHamiltonian, metric::AHMC.AbstractMetric; ϵ=alg.ϵ)
+    pc = AHMC.Preconditioner(metric)
+    da = AHMC.NesterovDualAveraging(alg.δ, ϵ)
+    if metric == AHMC.UnitEuclideanMetric
+        adaptor = AHMC.NaiveHMCAdaptor(pc, da)
     else
-        adaptor = AHMC.StanHMCAdaptor(alg.n_adapts, p, nda)
+        adaptor = AHMC.StanHMCAdaptor(alg.n_adapts, pc, da)
     end
     return adaptor
 end
 
+AHMCAdaptor(::Hamiltonian, ::AHMC.AbstractMetric; kwargs...) = AHMC.Adaptation.NoAdaptation()
+
+##########################
+# HMC State Constructors #
+##########################
+
+function HMCState(
+    model::Model,
+    spl::Sampler{<:Hamiltonian},
+    rng::AbstractRNG;
+    kwargs...
+)
+    # Reuse the VarInfo.
+    vi = spl.state.vi
+
+    # Link everything if needed.
+    !islinked(vi, spl) && link!(vi, spl)
+
+    # Get the initial log pdf and gradient functions.
+    ∂logπ∂θ = gen_∂logπ∂θ(vi, spl, model)
+    logπ = gen_logπ(vi, spl, model)
+
+    # Get the metric type.
+    metricT = getmetricT(spl.alg)
+
+    # Create a Hamiltonian.
+    θ_init = Vector{Float64}(spl.state.vi[spl])
+    metric = metricT(length(θ_init))
+    h = AHMC.Hamiltonian(metric, logπ, ∂logπ∂θ)
+
+    # Find good eps if not provided one
+    if spl.alg.ϵ == 0.0
+        ϵ = AHMC.find_good_eps(h, θ_init)
+        @info "Found initial step size" ϵ
+    else
+        ϵ = spl.alg.ϵ
+    end
+
+    # Generate a trajectory.
+    traj = gen_traj(spl.alg, ϵ)
+
+    # Generate a phasepoint. Replaced during sample_init!
+    h, t = AHMC.sample_init(rng, h, θ_init) # this also ensure AHMC has the same dim as θ.
+
+    # Unlink everything.
+    invlink!(vi, spl)
+
+    return HMCState(vi, 0, 0, traj, h, AHMCAdaptor(spl.alg, metric; ϵ=ϵ), t.z)
+end
+
+#######################################################
+# Special callback functionality for the HMC samplers #
+#######################################################
+
+mutable struct HMCCallback{
+    ProgType<:ProgressMeter.AbstractProgress
+} <: AbstractCallback
+    p :: ProgType
+end
+
+
+function callback(
+    rng::AbstractRNG,
+    model::ModelType,
+    spl::SamplerType,
+    N::Integer,
+    iteration::Integer,
+    t::HamiltonianTransition,
+    cb::HMCCallback;
+    kwargs...
```

Comment:
Do we need the `progress` keyword argument here?

Reply:
No, it's being handled at the `sample` level:

https://github.com/TuringLang/Turing.jl/blob/d992891b17c40506c93be61f493065e4dc905b76/src/interface/Interface.jl#L162

Reply:
The default callback still take progress. Maybe remove it?

Reply:
Oh, I see what you mean. Good catch. I removed the `kwarg` from the default `callback`.

Reply:
Great!